ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/commons/RaoUtil.java

```diff
@@ -104,4 +116,76 @@ public static double roundDouble(double value, int numberOfBitsToRoundOff) {
         }
         return value;
     }
+
+    /**
+     * Returns true if a remedial action is available depending on its usage rules
+     * If it has a OnFlowConstraint usage rule, then the margins are needed
+     */
+    public static boolean isRemedialActionAvailable(RemedialAction<?> remedialAction, State optimizedState, FlowResult flowResult) {
+        switch (remedialAction.getUsageMethod(optimizedState)) {
+            case AVAILABLE:
+                return true;
+            case TO_BE_EVALUATED:
+                return remedialAction.getUsageRules().stream()
+                    .anyMatch(usageRule -> (usageRule instanceof OnFlowConstraint)
+                        && isOnFlowConstraintAvailable((OnFlowConstraint) usageRule, optimizedState, flowResult));
+            default:
+                return false;
+        }
+    }
+
+    /**
+     * Returns true if a OnFlowConstraint usage rule is verified, ie if the associated CNEC has a negative margin
+     * It needs a FlowResult to get the margin of the flow cnec
+     */
+    public static boolean isOnFlowConstraintAvailable(OnFlowConstraint onFlowConstraint, State optimizedState, FlowResult flowResult) {
+        if (!onFlowConstraint.getUsageMethod(optimizedState).equals(UsageMethod.TO_BE_EVALUATED)) {
+            return false;
+        } else {
+            // We don't actually need to know the unit of the objective function, we just need to know if the margin is negative
+            return flowResult.getMargin(onFlowConstraint.getFlowCnec(), Unit.MEGAWATT) <= 0;
+        }
+    }
+
+    public static Pair<RangeAction<?>, State> getLastAvailableRangeActionOnSameAction(OptimizationPerimeter optimizationContext, RangeAction<?> rangeAction, State state) {
+
+        if (state.isPreventive() || state.equals(optimizationContext.getMainOptimizationState())) {
+            // no previous instant
+            return null;
+        } else if (state.getInstant().equals(Instant.CURATIVE)) {
+
+            // look if a preventive range action acts on the same network elements
+            State preventiveState = optimizationContext.getRangeActionsPerState().keySet().stream().filter(State::isPreventive).findAny().orElse(null);
```

Comment:
an entry is actually a pair<State, Set<RangeAction>>
I haven't found a clean way of doing it in one pass but looking at the method, I believe instead of looking for the preventive state we should be able to use the mainOptimizationState of the optimizationContext.

I'm not a huge fan of this method though, it seems very specific to cases where we just have a preventive and curative states... However this does allow the method  to be quite efficient (even more efficient if we dont look for the preventive state but use the mainOptimizationState)

Reply:
you do not need to store the "preventiveState" variable then, it will be even clearer to read